client/components/notice/index.tsx

```diff
@@ -0,0 +1,146 @@
+/**
+ * External dependencies
+ */
+
+import React, { useEffect, FunctionComponent, useRef } from 'react';
+import classnames from 'classnames';
+import { localize } from 'i18n-calypso';
+import Gridicon from 'calypso/components/gridicon';
+
+/**
+ * Internal dependencies
+ */
+import { TimerHandle } from 'calypso/types';
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+/**
+ * Module constants
+ */
+const GRIDICONS_WITH_DROP = [
+	'add',
+	'cross-circle',
+	'ellipsis-circle',
+	'help',
+	'info',
+	'notice',
+	'pause',
+	'play',
+	'spam',
+];
+
+/**
+ * Constants
+ */
+enum Status {
+	Error = 'is-error',
+	Info = 'is-info',
+	Success = 'is-success',
+	Warning = 'is-warning',
+	Plain = 'is-plain',
+}
+
+interface Props {
+	className?: string;
+	duration?: number;
+	onDismissClick: () => void;
+	translate: ( x: string ) => string;
+	children: React.ReactNode;
+	status?: Status;
+	icon?: string | null;
+	isCompact?: boolean;
+	isLoading?: boolean;
+	showDismiss?: boolean;
+	text?: string | React.ReactNode;
+}
+
+const getIcon = ( status: Status ): string => {
+	let icon: string;
+
+	switch ( status ) {
+		case Status.Info:
+			icon = 'info';
+			break;
+		case Status.Success:
+			icon = 'checkmark';
+			break;
+		case Status.Error:
+			icon = 'notice';
+			break;
+		case Status.Warning:
+			icon = 'notice';
+			break;
+		default:
+			icon = 'info';
+			break;
+	}
+
+	return icon;
+};
+
+const Notice: FunctionComponent< Props > = ( {
+	className = '',
+	duration = 0,
+	onDismissClick,
+	translate,
+	children,
+	status,
+	icon,
+	isCompact = false,
+	isLoading = false,
+	showDismiss = ! isCompact, // by default, show on normal notices, don't show on compact ones
+	text,
+} ) => {
+	const dismissTimeout = useRef< TimerHandle | null >( null );
+
+	useEffect( () => {
+		if ( dismissTimeout.current ) {
+			clearTimeout( dismissTimeout.current );
+		}
+
+		if ( duration > 0 ) {
+			dismissTimeout.current = setTimeout( onDismissClick, duration );
```

Comment:
Should we be doing these only if `onDismissClick` is defined and not a noop function?

Reply:
The `dismissTimeout` doesn't need to be stored in a ref. It can be a local variable in the `useEffect` callback scope.